cpp/src/sort/rank.cu

```diff
@@ -0,0 +1,220 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cudf/column/column.hpp>
+#include <cudf/column/column_factories.hpp>
+#include <cudf/detail/sorting.hpp>
+#include <cudf/sorting.hpp>
+#include <cudf/table/row_operators.cuh>
+#include <cudf/table/table.hpp>
+#include <cudf/table/table_device_view.cuh>
+#include <cudf/table/table_view.hpp>
+#include <cudf/utilities/error.hpp>
+
+#include <rmm/thrust_rmm_allocator.h>
+#include <thrust/iterator/discard_iterator.h>
+#include <thrust/sequence.h>
+
+namespace cudf {
+namespace experimental {
+namespace detail {
+
+template <bool has_nulls, typename ReturnType = bool>
+struct unique_comparator {
+  unique_comparator(table_device_view device_table,
+                    size_type const *sorted_order)
+      : comp(device_table, device_table, true), perm(sorted_order) {}
+  __device__ ReturnType operator()(size_type index) const noexcept {
+    return index == 0 || not comp(perm[index], perm[index - 1]);
+  };
+
+private:
+  row_equality_comparator<has_nulls> comp;
+  size_type const *perm;
+};
+
+std::unique_ptr<table> rank(table_view const &input, rank_method method,
+                            order column_order, include_nulls _include_nulls,
+                            null_order null_precedence,
+                            bool percentage,
+                            rmm::mr::device_memory_resource *mr,
+                            cudaStream_t stream = 0) {
+  auto const size = input.num_rows();
+
+  std::vector<std::unique_ptr<column>> rank_columns;
+  for (auto const &input_col : input) {
+    std::unique_ptr<column> sorted_order =
+        (method == rank_method::FIRST)
+            ? detail::stable_sorted_order(table_view{{input_col}},
+                                          {column_order}, {null_precedence}, mr,
+                                          stream)
+            : detail::sorted_order(table_view{{input_col}}, {column_order},
+                                   {null_precedence}, mr, stream);
+    column_view sorted_order_view = sorted_order->view();
+
+    // na_option=keep assign NA to NA values
+    if (_include_nulls == include_nulls::NO)
+      rank_columns.push_back(make_numeric_column(
+          data_type(FLOAT64), size, copy_bitmask(input_col, stream, mr),
+          input_col.null_count(), stream, mr));
+    else
+      rank_columns.push_back(make_numeric_column(
+          data_type(FLOAT64), size, mask_state::UNALLOCATED, stream, mr));
+
+    auto rank_mutable_view = rank_columns.back()->mutable_view();
+    auto rank_data = rank_mutable_view.data<double>();
+    auto device_table =
+        table_device_view::create(table_view{{input_col}}, stream);
+
+    // All equal values have same rank and rank always increases by 1 between
+    // groups
+    rmm::device_vector<size_type> dense_rank_sorted; // as key for min, max,
```

Comment:
I think this is a good example of @codereport's "Initial Then Modify" anti-pattern. This looks like a good case for an IILE.

Reply:
Also, since this isn't used in the `rank_method::FIRST` case, it would be a lot clearer if this were a function that was called in the other cases.

Reply:
`dense_rank_sorted` is again used `pct=True` and `rank_method::DENSE`. so, I couldn't make it a local variable to each case.